src/transformers/generation/tf_utils.py

```diff
@@ -3066,16 +3066,20 @@ def beam_search_cond_fn(
             not_max_length_yet = cur_len < max_length
 
             # 2. can the new beams still improve?
-            best_running_score = running_scores[:, :1] / (max_length**length_penalty)
+            best_running_score = running_scores[:, :1] / tf.cast(cur_len, dtype=running_scores.dtype) ** length_penalty
```

Comment:
In current `main` branch, `max_length` is used instead of `cur_len`. However, in our PyTorch generation's `BeamHypotheses`, it is `cur_len`, see 
https://github.com/huggingface/transformers/blob/3be028bc9d4b2cce9539b940f17052f333284684/src/transformers/generation/beam_search.py#L895

When running the code snippet in the reported [`TFMarian` issue (https://github.com/huggingface/transformers/issues/18149), we get `max_length` being a constant of `512`, but the PyTorch generation code runs with `cur_len` which is from `1` (or `2`) to `5`.

(However, this is not the root cause of the issue in #18149)